### PR TITLE
client/{core, webserver, eth}: Manual token approvals

### PR DESCRIPTION
**Pull request by martonp**
This makes token approvals a separate step for users instead of an automatic part of the first token swap. This allows users to be aware of what is happening and not be suprised by the high cost of their swap.

If the base asset is an unnapproved token, then the user will not be able to to a Sell. The Sell button in the market page will be disabled, and there will be a button which allows the user to do the approve transaction. If the quote asset is an unnapproved token, then the Buy button will be disabled. If both assets are unnaproved tokens, the order panel will not be displayed.
Screenshot 2023-05-10 at 10 24 29 AM

Screenshot 2023-05-10 at 10 24 50 AM
Screenshot 2023-05-10 at 10 25 21 AM

### CODE REVIEW

File: client/asset/eth/eth.go
Comment (martonp) [reply client/asset/eth/eth.go:2208]: True, I guess non-approval could also be cached since the wallet will know when an approval is done.

File: client/asset/eth/eth.go
Comment (martonp) [reply client/asset/eth/eth.go:518]: I'll just add it in this PR.

**Comment by martonp** [comment]
Added functionality to unnapprove:

Screenshot 2023-05-13 at 8 43 20 AM
Screenshot 2023-05-13 at 8 43 31 AM

**Comment by chappjc** [comment]
This looks A-OK, but I got stuck with the dcr harness (https://github.com/decred/dcrdex/pull/2358).  Will give it a final test and merge today.

File: client/webserver/site/src/html/markets.tmpl
Comment (chappjc) [inline client/webserver/site/src/html/markets.tmpl:209]: The "Approval Pending" will have to go in the locales, either backend html renderer or the javascript.

I also noticed that after unapproving, it said "Approval Pending", which is not wrong since it's approving 0, but could be confusing.  Maybe "Approval Change Pending"?

**Review by chappjc — CHANGES_REQUESTED** [review]
Working well.  Two things I commented on though:

- wording and localization of one string on the frontend
- some thoughts about how we could make things work for existing headless consumers with no need for awareness of tokens or approvals at the application level, which could be a bot or third-party code we don't want to break.

File: client/core/core.go
Comment (chappjc) [reply client/core/core.go:5375]: I guess `WalletState` covers the above. nvm
But we still might want to consider some of the stuff in https://github.com/decred/dcrdex/pull/2349#discussion_r1196935906

File: client/asset/eth/eth.go
Comment (martonp) [reply client/asset/eth/eth.go:1449]: I agree that a new error type is warranted. If clients are already doing special error handling based on the new error type, then might as well have them also do an `ApproveToken` call I think. I'm not totally against it, but it feels kind of weird that `FundOrder` spends money without it having been approved by the caller. The split transaction spends money, but that was at least approved by the wallet settings or the order config.

I'm thinking it would be helpful to add an `onConfirm func()` argument to the `ApproveToken` function, then clients can easily know when it's ready to trade.

Comment (chappjc) [reply client/asset/eth/eth.go:1449]: Good points.  Clients will just have to adapt to this change and call the approve functions themselves.

**Review by chappjc — APPROVED** [review]

**Pull request merged**
